src/fromager/bootstrapper.py

```diff
@@ -0,0 +1,489 @@
+    def __init__(
+        self,
+        ctx: context.WorkContext,
+        progressbar: progress.Progressbar | None = None,
+        prev_graph: DependencyGraph | None = None,
+    ) -> None:
+        self.ctx = ctx
+        self.progressbar = progressbar or progress.Progressbar(None)
+        self.prev_graph = prev_graph
+        self.why: list[tuple[RequirementType, Requirement, Version]] = []
+
+    def bootstrap(self, req: Requirement, req_type: RequirementType) -> Version:
+        constraint = self.ctx.constraints.get_constraint(req.name)
+        if constraint:
+            logger.info(
+                f"{req.name}: incoming requirement {req} matches constraint {constraint}. Will apply both."
+            )
+
+        pbi = self.ctx.package_build_info(req)
+        if pbi.pre_built:
+            resolved_version, wheel_url, wheel_filename, unpack_dir = (
+                self._handle_prebuilt_sources(req, req_type)
+            )
+            # Remember that this is a prebuilt wheel, and where we got it.
+            source_url = wheel_url
+            source_url_type = "prebuilt"
+        else:
+            resolved_version, source_url, source_filename, source_url_type = (
+                self._handle_sources(req, req_type)
+            )
+
+        self._add_to_graph(req, req_type, resolved_version, source_url)
+
+        # Avoid cyclic dependencies and redundant processing.
+        if self.ctx.has_been_seen(req, resolved_version):
+            logger.debug(
+                f"{req.name}: redundant {req_type} requirement {self.why} -> {req} resolves to {resolved_version}"
+            )
+            return resolved_version
+        self.ctx.mark_as_seen(req, resolved_version)
+
+        logger.info(
+            f"{req.name}: new {req_type} dependency {req} resolves to {resolved_version}"
+        )
+
+        # Build the dependency chain up to the point of this new
+        # requirement using a new list so we can avoid modifying the list
+        # we're given.
+        self.why.append((req_type, req, resolved_version))
+
+        # for cleanup
+        build_env = None
+        sdist_root_dir = None
+        if not pbi.pre_built:
+            (
+                sdist_root_dir,
+                unpack_dir,
+                build_dependencies,
+            ) = self._prepare_source(req, source_filename, resolved_version)
+
+            found_wheel_filename = self._is_wheel_built(req, resolved_version)
+            if not found_wheel_filename:
+                wheel_filename, build_env = self._build(
+                    req, resolved_version, sdist_root_dir, build_dependencies
+                )
+            else:
+                wheel_filename = found_wheel_filename
+
+        self.ctx.add_to_build_order(
+            req=req,
+            version=resolved_version,
+            source_url=source_url,
+            source_url_type=source_url_type,
+            prebuilt=pbi.pre_built,
+            constraint=constraint,
+        )
```

Comment:
I moved this to after building of wheels because:
- We have graph.json that as already recorded this dependency
- if the build fails then anyways bootstrapping stops and we don't end up using that build order file